libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/roadname/view/MapboxRoadNameView.kt

```diff
@@ -0,0 +1,150 @@
+import android.graphics.Bitmap
+import android.text.SpannableStringBuilder
+import android.text.Spanned
+import android.text.style.ImageSpan
+import android.util.AttributeSet
+import android.util.Log
+import androidx.appcompat.widget.AppCompatTextView
+import com.mapbox.bindgen.Expected
+import com.mapbox.navigation.base.road.model.Road
+import com.mapbox.navigation.ui.maps.roadname.api.MapboxRoadNameLabelApi
+import com.mapbox.navigation.ui.shield.api.toBitmap
+import com.mapbox.navigation.ui.shield.model.RouteShield
+import com.mapbox.navigation.ui.shield.model.RouteShieldError
+import com.mapbox.navigation.ui.shield.model.RouteShieldResult
+import com.mapbox.navigation.ui.utils.internal.extensions.drawableWithHeight
+import com.mapbox.navigation.ui.utils.internal.ifNonNull
+
+/**
+ * Default Mapbox implementation that allows you to render road name labels and route shields
+ * associated with the name.
+ */
+class MapboxRoadNameView : AppCompatTextView {
+
+    private var shieldName: String? = null
+    private var currentRoadName: String? = null
+    private val shields = mutableSetOf<RouteShieldResult>()
+
+    /**
+     * Default view to render a road name label.
+     *
+     * @see MapboxRoadNameLabelApi
+     */
+    constructor(context: Context) : super(context)
+
+    /**
+     * Default view to render a road name label.
+     *
+     * @see MapboxRoadNameLabelApi
+     */
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
+
+    /**
+     * Default view to render a road name label.
+     *
+     * @see MapboxRoadNameLabelApi
+     */
+    constructor(
+        context: Context,
+        attrs: AttributeSet?,
+        defStyleAttr: Int
+    ) : super(context, attrs, defStyleAttr)
+
+    /**
+     * Invoke the method to render the view with road name.
+     * @param road Road
+     */
+    fun renderRoadName(road: Road) {
+        shieldName = road.shieldName
+        currentRoadName = road.name
+        renderRoadNameLabel()
+    }
+
+    /**
+     * Invoke the method to render the view with road shields.
+     * @param shields list of shields to render
+     */
+    fun renderRoadNameWith(shields: List<Expected<RouteShieldError, RouteShieldResult>>) {
+        shields.forEach { result ->
+            result.fold(
+                {
+                    Log.d("abhishek_testing", "$it")
+                },
+                { routeShieldResult ->
+                    this.shields.add(routeShieldResult)
+                }
+            )
+        }
+        renderRoadNameLabel()
+    }
+
+    private fun renderRoadNameLabel() {
+        val roadNameLabel = SpannableStringBuilder()
+        shields.forEach { routeShieldResult ->
+            val shieldBitmap = when (routeShieldResult.shield) {
+                is RouteShield.MapboxDesignedShield -> {
+                    (routeShieldResult.shield as RouteShield.MapboxDesignedShield)
+                        .toBitmap(context.resources, lineHeight)
```

Comment:
This method is the same for both types. Why not move it from the Extension to `RouteShield` (and make it abstract for both types to implement)?

We could then call it without a cast and it would be easier to consume for Java users.

Reply:
Resolved.